pkg/chunk/aws/dynamodb_index_reader.go

```diff
@@ -0,0 +1,238 @@
+}
+
+// NewDynamoDBIndexReader returns an object that can scan an entire index table
+func NewDynamoDBIndexReader(cfg DynamoDBConfig, schemaCfg chunk.SchemaConfig, reg prometheus.Registerer, l gklog.Logger, rowsRead prometheus.Counter) (chunk.IndexReader, error) {
+	client, err := newDynamoDBStorageClient(cfg, schemaCfg, reg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dynamodbIndexReader{
+		dynamoDBStorageClient: *client,
+		maxRetries:            cfg.BackoffConfig.MaxRetries,
+		log:                   l,
+
+		rowsRead: rowsRead,
+	}, nil
+}
+
+func (r *dynamodbIndexReader) IndexTableNames(ctx context.Context) ([]string, error) {
+	// fake up a table client - if we call NewDynamoDBTableClient() it will double-register metrics
+	tableClient := dynamoTableClient{
+		DynamoDB: r.DynamoDB,
+		metrics:  r.metrics,
+	}
+	return tableClient.ListTables(ctx)
+}
+
+type seriesMap struct {
+	mutex           sync.Mutex           // protect concurrent access to maps
+	seriesProcessed map[string]sha256Set // map of userID/bucket to set showing which series have been processed
+}
+
+// Since all sha256 values are the same size, a fixed-size array
+// is more space-efficient than string or byte slice
+type sha256 [32]byte
+
+// an entry in this set indicates we have processed a series with that sha already
+type sha256Set struct {
+	series map[sha256]struct{}
+}
+
+// ReadIndexEntries reads the whole of a table on multiple goroutines in parallel.
+// Entries for the same HashValue and RangeValue should be passed to the same processor.
+func (r *dynamodbIndexReader) ReadIndexEntries(ctx context.Context, tableName string, processors []chunk.IndexEntryProcessor) error {
+	var outerErr error
+	projection := hashKey + "," + rangeKey
+
+	sm := &seriesMap{ // new map per table
+		seriesProcessed: make(map[string]sha256Set),
+	}
+
+	var readerGroup sync.WaitGroup
+	readerGroup.Add(len(processors))
+	// Start a goroutine for each processor
+	for i, processor := range processors {
+		go func(segment int, processor chunk.IndexEntryProcessor) {
+			input := &dynamodb.ScanInput{
+				TableName:              aws.String(tableName),
+				ProjectionExpression:   aws.String(projection),
+				Segment:                aws.Int64(int64(segment)),
+				TotalSegments:          aws.Int64(int64(len(processors))),
+				ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+			}
+			withRetrys := func(req *request.Request) {
+				req.Retryer = client.DefaultRetryer{NumMaxRetries: r.maxRetries}
+			}
+			err := r.DynamoDB.ScanPagesWithContext(ctx, input, func(page *dynamodb.ScanOutput, lastPage bool) bool {
+				if cc := page.ConsumedCapacity; cc != nil {
+					r.metrics.dynamoConsumedCapacity.WithLabelValues("DynamoDB.ScanTable", *cc.TableName).
+						Add(float64(*cc.CapacityUnits))
+				}
+				r.processPage(ctx, sm, processor, tableName, page)
+				return true
+			}, withRetrys)
+			if err != nil {
+				outerErr = err
+				// TODO: abort all segments
```

Comment:
I think this is a good case for `errgroup.WithContext()` (we already use it in some places). You can create an `errgroup`, spawn the goroutines with `group.Go()` and the whole group context would be canceled as soon as the first gorouting returns an error. Then the error will be returned to `group.Wait()`.